### PR TITLE
Audio functions

### PR DESCRIPTION
**Pull request by harisreedhar**

### CODE REVIEW

File: facefusion/ffmpeg.py [hunk 1]
```diff
@@ -8,91 +8,107 @@
...
 def run_ffmpeg(args : List[str]) -> bool:
-	commands = [ 'ffmpeg', '-hide_banner', '-loglevel', 'error' ]
-	commands.extend(args)
-	try:
-		subprocess.run(commands, stderr = subprocess.PIPE, check = True)
-		return True
-	except subprocess.CalledProcessError as exception:
-		logger.debug(exception.stderr.decode().strip(), __name__.upper())
-		return False
+    commands = [ 'ffmpeg', '-hide_banner', '-loglevel', 'error' ]
```
Comment (henryruhs) [inline facefusion/ffmpeg.py:11]: Undo replacing tabs with spaces

File: facefusion/ffmpeg.py [hunk 2]
```diff
@@ -80,6 +80,22 @@ def restore_audio(target_path : str, output_path : str, video_fps : Fps) -> bool
 	return run_ffmpeg(commands)
 
 
+def read_audio_buffer(target_path : str, sample_rate : int, channel_total : int) -> Optional[bytes]:
+	commands = [ '-i', target_path, '-vn', '-f', 's16le', '-acodec', 'pcm_s16le', '-ar', str(sample_rate), '-ac', str(channel_total), '-' ]
+	process = open_ffmpeg(commands)
+	audio_buffer, error = process.communicate()
+	if process.returncode != 0:
```
Comment (henryruhs) [inline facefusion/ffmpeg.py:87]: Flip that one...

```
if process.returncode == 0:
		return audio_buffer
logger.debug(error.decode().strip(), __name__.upper())
return None
````

File: facefusion/audio.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+from typing import Optional, Any, List
+
+import numpy
+import scipy
+from functools import lru_cache
+from facefusion.ffmpeg import read_audio_buffer
+from facefusion.typing import Fps
+
+
+def monoize_audio(audio : numpy.ndarray[Any, Any]) -> numpy.ndarray[Any, Any]:
```
Comment (henryruhs) [inline facefusion/audio.py:10]: normalize_audio() ?

File: facefusion/audio.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+def cleanse_audio(audio : numpy.ndarray[Any, Any]) -> numpy.ndarray[Any, Any]:
+	audio = audio / numpy.max(numpy.abs(audio), axis = 0)
+	audio = scipy.signal.resample(audio, len(audio))
+	audio = scipy.signal.lfilter([1, -0.97], [1], audio)
+	return audio
+
+
+@lru_cache(maxsize = None)
+def create_static_mel_filter(sample_rate : int, filter_total : int, filter_size : int) -> numpy.ndarray[Any, Any]:
+	convert_mel_to_hz = lambda x: 700 * (10 ** (x / 2595) - 1)
```
Comment (henryruhs) [inline facefusion/audio.py:25]: Move the lambda to their own def blocks

File: facefusion/audio.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+from facefusion.typing import Fps
+
+
+def monoize_audio(audio : numpy.ndarray[Any, Any]) -> numpy.ndarray[Any, Any]:
+	if audio.ndim > 1:
+		audio = numpy.mean(audio, axis = 1)
+	return audio
+
+
+def cleanse_audio(audio : numpy.ndarray[Any, Any]) -> numpy.ndarray[Any, Any]:
```
Comment (henryruhs) [inline facefusion/audio.py:16]: cleanse? or clean_audio? what do you clean it from?

numpy.ndarray[Any, Any]) -> numpy.ndarray[Any, Any]:

please create alias types that give a better audio what is being processed

File: facefusion/ffmpeg.py [hunk 6]
```diff
@@ -80,6 +80,22 @@ def restore_audio(target_path : str, output_path : str, video_fps : Fps) -> bool
 	return run_ffmpeg(commands)
 
 
+def read_audio_buffer(target_path : str, sample_rate : int, channel_total : int) -> Optional[bytes]:
```
Comment (henryruhs) [inline facefusion/ffmpeg.py:83]: Return type AudioBuffer

File: facefusion/audio.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	audio = numpy.frombuffer(buffer, dtype = numpy.int16).reshape(-1, 2)
+	audio = monoize_audio(audio)
+	audio = cleanse_audio(audio)
+	spectrogram = create_spectrogram(audio, 16000, 80, 800)
+	audio_frames = extract_spectrogram(spectrogram, 80, 16, fps)
+	return audio_frames
+
+
+def get_audio_frame(audio_path : str, fps : Fps, frame_number : int) -> Optional[numpy.ndarray[Any, Any]]:
+	if audio_path:
```
Comment (henryruhs) [inline facefusion/audio.py:67]: Move the method to the very top of the file

Is audio frame a real word?

`Optional[numpy.ndarray[Any, Any]]:` ... hard to understand what this actual returns

File: facefusion/audio.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@lru_cache(maxsize = None)
+def read_static_audio(audio_path : str, fps : Fps) -> List[numpy.ndarray[Any, Any]]:
+	buffer = read_audio_buffer(audio_path, 16000, 2)
+	audio = numpy.frombuffer(buffer, dtype = numpy.int16).reshape(-1, 2)
+	audio = monoize_audio(audio)
+	audio = cleanse_audio(audio)
+	spectrogram = create_spectrogram(audio, 16000, 80, 800)
+	audio_frames = extract_spectrogram(spectrogram, 80, 16, fps)
```
Comment (henryruhs) [inline facefusion/audio.py:62]: audio_frames = extract_spectrogram <- naming missmatch, extract_audio_frames() ?

File: facefusion/audio.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+			audio_frames.append(spectrogram[:, spectrogram.shape[1] - steps : ])
+			break
+		audio_frames.append(spectrogram[:, start_index : start_index + steps])
+		index += 1
+	return audio_frames
+
+
+@lru_cache(maxsize = None)
+def read_static_audio(audio_path : str, fps : Fps) -> List[numpy.ndarray[Any, Any]]:
+	buffer = read_audio_buffer(audio_path, 16000, 2)
```
Comment (henryruhs) [inline facefusion/audio.py:57]: rename to audio_buffer

File: facefusion/audio.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@lru_cache(maxsize = None)
+def create_static_mel_filter(sample_rate : int, filter_total : int, filter_size : int) -> numpy.ndarray[Any, Any]:
+	convert_mel_to_hz = lambda x: 700 * (10 ** (x / 2595) - 1)
+	convert_hz_to_mel = lambda x: 2595 * numpy.log10(1 + x / 700)
+	bins = numpy.linspace(convert_hz_to_mel(55), convert_hz_to_mel(7600), filter_total + 2) # type: ignore
+	indices = numpy.floor((filter_size + 1) * convert_mel_to_hz(bins) / sample_rate).astype(numpy.int32) # type: ignore
+	filters = numpy.zeros((filter_total, filter_size // 2 + 1))
+	for i in range(1, filter_total + 1):
```
Comment (henryruhs) [inline facefusion/audio.py:30]: rename i to index...

File: facefusion/audio.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+	mel_filters = create_static_mel_filter(sample_rate, filter_total, filter_size)
+	spectrogram = scipy.signal.stft(audio, nperseg = filter_size, noverlap = 600, nfft = filter_size)[2]
+	spectrogram = numpy.dot(mel_filters, numpy.abs(spectrogram))
+	return spectrogram
+
+
+def extract_spectrogram(spectrogram : numpy.ndarray[Any, Any], filter_total : int, steps : int, fps : float) -> List[numpy.ndarray[Any, Any]]:
+	audio_frames = []
+	index = 0
+	while True:
```
Comment (henryruhs) [inline facefusion/audio.py:45]: Ask ChatGPT - there must be a smarter way to use a while True that has a loop and break inside :D

Comment (harisreedhar) [reply facefusion/audio.py:57]: on future update

File: facefusion/audio.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+		audio = numpy.mean(audio, axis = 1)
+	audio = audio / numpy.max(numpy.abs(audio), axis = 0)
+	audio = scipy.signal.resample(audio, len(audio))
+	audio = scipy.signal.lfilter([1, -0.97], [1], audio)
+	return audio
+
+
+@lru_cache(maxsize = None)
+def create_static_mel_filter(sample_rate : int, filter_total : int, filter_size : int) -> numpy.ndarray[Any, Any]:
+	bins = (10 ** (numpy.linspace(85, 2787, filter_total + 2) / 2595) - 1) * 700
```
Comment (henryruhs) [inline facefusion/audio.py:39]: we cannot leave it that way - this are kinda magic numbers that only you understand.

but neither me or your future self do :D

File: facefusion/audio.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+	audio_frames = extract_audio_frames(spectrogram, 80, 16, fps)
+	return audio_frames
+
+
+def normalize_audio(audio : numpy.ndarray[Any, Any]) -> Audio:
+	if audio.ndim > 1:
+		audio = numpy.mean(audio, axis = 1)
+	audio = audio / numpy.max(numpy.abs(audio), axis = 0)
+	audio = scipy.signal.resample(audio, len(audio))
+	audio = scipy.signal.lfilter([1, -0.97], [1], audio)
```
Comment (henryruhs) [inline facefusion/audio.py:33]: may I ask where the -0.97 comes from? why not -0.95 or -0.98

File: facefusion/audio.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+def get_audio_frame(audio_path : str, fps : Fps, frame_number : int) -> Optional[AudioFrame]:
+	if audio_path:
+		audio_frames = read_static_audio(audio_path, fps)
+		if frame_number < len(audio_frames):
+			return audio_frames[frame_number]
+	return None
+
+
+@lru_cache(maxsize = None)
+def read_static_audio(audio_path : str, fps : Fps) -> List[Optional[AudioFrame]]:
```
Comment (henryruhs) [inline facefusion/audio.py:19]: I prefer to have `List[AudioFrame]:` ... can we filter the empty entries?

File: facefusion/audio.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
+from typing import Optional, Any, List
+
+import numpy
+import scipy
+from functools import lru_cache
+from facefusion.ffmpeg import read_audio_buffer
+from facefusion.typing import Fps, Audio, Spectrogram, AudioFrame
+
+
+def get_audio_frame(audio_path : str, fps : Fps, frame_number : int) -> Optional[AudioFrame]:
```
Comment (henryruhs) [inline facefusion/audio.py:10]: I suggest `frame_number : int = 0` like get_video_frame()

File: facefusion/audio.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	bins = (10 ** (numpy.linspace(85, 2787, filter_total + 2) / 2595) - 1) * 700
+	indices = numpy.floor((filter_size + 1) * bins / sample_rate).astype(numpy.int16)
+	filters = numpy.zeros((filter_total, filter_size // 2 + 1))
+	for index in range(1, filter_total + 1):
+		filters[index - 1, indices[index - 1] : indices[index]] = scipy.signal.windows.triang(indices[index] - indices[index - 1])
+	return filters
+
+
+def create_spectrogram(audio : Audio, sample_rate : int, filter_total : int, filter_size : int) -> Spectrogram:
+	mel_filters = create_static_mel_filter(sample_rate, filter_total, filter_size)
```
Comment (henryruhs) [inline facefusion/audio.py:48]: mel_filters = create_static_mel_filter ... either the variable or the method are wrong (plural vs. singular)

File: facefusion/audio.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+		if frame_number < len(audio_frames):
+			return audio_frames[frame_number]
+	return None
+
+
+@lru_cache(maxsize = None)
+def read_static_audio(audio_path : str, fps : Fps) -> List[Optional[AudioFrame]]:
+	audio_buffer = read_audio_buffer(audio_path, 16000, 2)
+	audio = numpy.frombuffer(audio_buffer, dtype = numpy.int16).reshape(-1, 2)
+	audio = normalize_audio(audio)
```
Comment (henryruhs) [inline facefusion/audio.py:22]: this coding style is great - having one method for each task 👍🏻

File: facefusion/audio.py (hunk 12 above)
Comment (harisreedhar) [reply facefusion/audio.py]: i managed to move those numbers inside `read_static_audio()`.
what do you suggest for storing the values? A global dict like this?
```python
AudioParameters : Dict[Any] = {
    ''sample_rate": 16000,
    "frequency_minimum": 55.0,
    "frequency_maximum": 7600.0,
    "filter_total": 80,
    "filter_size": 800,
    "filter_coefficient: -0.97"
}
```

File: facefusion/audio.py (hunk 13 above)
Comment (harisreedhar) [reply facefusion/audio.py]: https://github.com/facefusion/facefusion/pull/345#discussion_r1468759873

File: facefusion/ffmpeg.py [hunk 18]
```diff
@@ -80,6 +80,22 @@ def restore_audio(target_path : str, output_path : str, video_fps : Fps) -> bool
 	return run_ffmpeg(commands)
 
 
+def read_audio_buffer(target_path : str, sample_rate : int, channel_total : int) -> Optional[AudioBuffer]:
+	commands = [ '-i', target_path, '-vn', '-f', 's16le', '-acodec', 'pcm_s16le', '-ar', str(sample_rate), '-ac', str(channel_total), '-' ]
+	process = open_ffmpeg(commands)
+	audio_buffer, error = process.communicate()
+	if process.returncode == 0:
+		return audio_buffer
+	logger.debug(error.decode().strip(), __name__.upper())
```
Comment (henryruhs) [inline facefusion/ffmpeg.py:89]: can we move that to open_ffmpeg() ?

File: facefusion/audio.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def convert_mel_to_hertz(mel : numpy.ndarray[Any, Any]) -> numpy.ndarray[Any, Any]:
+	return 700 * (10 ** (mel / 2595) - 1)
+
+
+@lru_cache(maxsize = None)
+def create_static_mel_filter(sample_rate : int, filter_total : int, filter_size : int, frequency_minimum : float, frequency_maximum : float) -> numpy.ndarray[Any, Any]:
+	mel_filter = numpy.zeros((filter_total, filter_size // 2 + 1))
+	mel_bins = numpy.linspace(convert_hertz_to_mel(frequency_minimum), convert_hertz_to_mel(frequency_maximum), filter_total + 2)
```
Comment (henryruhs) [inline facefusion/audio.py:52]: I suggest two variables for

convert_hertz_to_mel(frequency_minimum)
convert_hertz_to_mel(frequency_maximum)

**Pull request merged**